pkg/bitcoin/hd.go

```diff
@@ -0,0 +1,58 @@
+package bitcoin
+
+import (
+	"context"
+
+	"github.com/btcsuite/btcutil/hdkeychain"
+	pb "github.com/ledgerhq/lama-bitcoin-svc/pb/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// References:
+//   [BIP32]: BIP0032 - Hierarchical Deterministic Wallets
+//   https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
+
+// DeriveExtendedKey provides an API to derive hierarchical deterministic
+// extended keys.
+//
+// There are no restrictions on the extended keys that can be derived, as
+// long as BIP0032 rules are followed. However, it is intended to be used
+// for deriving child keys from public extended keys at the account level
+// (HD depth 3).
+//
+// The derivation is agnostic of chain parameters. Derived extended keys
+// are associated to the same network as the parent extended key.
+//
+// The method's response includes the following fields:
+//     ExtendedKey: extended key as a human-readable base58-encoded string.
+//     PublicKey:   33-byte compressed public key of the derived extended key.
+//     ChainCode:   32-byte chain code of the derived extended key.
+func (s *Service) DeriveExtendedKey(
+	ctx context.Context, request *pb.DeriveExtendedKeyRequest,
+) (*pb.DeriveExtendedKeyResponse, error) {
+	extendedKey, err := hdkeychain.NewKeyFromString(request.ExtendedKey)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+
+	// Derive len(request.Derivation) HD levels, starting from extendedKey
+	// as the parent node.
+	for _, childIndex := range request.Derivation {
+		extendedKey, err = extendedKey.Child(childIndex)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, err.Error())
+		}
+	}
+
+	pubKey, err := extendedKey.ECPubKey()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+
+	return &pb.DeriveExtendedKeyResponse{
```

Comment:
Globally what is the guideline you are using for passing by values / reference choice ?

Reply:
Don't know how go handle this but better return by value. You don't know what happens to the address once the function goes out of scope.

Reply:
The function signature is generated by the Protobuf compiler, so not using pointer-to-struct is not an option if we want to implement the gRPC interface for Bitcoin.

Additionally, one demerit of using a pointer value here is that it'll prevent us from returning `nil` values in case of an error. We'll need to return a zero-value struct in all error cases, thereby increasing the allocations.

Reply:
> Globally what is the guideline you are using for passing by values / reference choice ?

I think your guideline is pretty good. However, protobuf structs cannot be considered light due to a lot of internal fields which is generally not used.

I like to use pointers when I don't need to care about the return value, such as the response of a service method. If this were a helper method, I wouldn't have used a pointer for sure.

Reply:
Clear, thank you @onyb 👍 
Actually my experience from proto usage was with singleton EMPTY value provided by generated class. Vut yes in this case if there is not sugar like this, pointer is justified, furthermore because this is pushed by protobuf itself.

Interesting benchmark on pointer usage : https://medium.com/@philpearl/bad-go-pointer-returns-340f2da8289

Reply:
> my experience from proto usage was with singleton EMPTY value provided by generated class
That's one killer feature! I'll see if this exists in the Go plugin for protobuf (for a future PR).

Thanks for the weekend read. Looks interesting! 👍